subsys/net/lib/download_client/src/download_client.c

```diff
@@ -791,6 +803,172 @@ static int handle_disconnect(struct download_client *client)
 	return err;
 }
 
+#if defined(CONFIG_DOWNLOAD_CLIENT_COAP_CLIENT)
+static void coap_cb(int16_t result_code, size_t offset, const uint8_t *payload, size_t len,
+			    bool last_block, void *user_data)
+{
+	int ret;
+	struct download_client *dl = (struct download_client *)user_data;
+	struct download_client_evt evt = {0};
+
+	LOG_DBG("CoAP client:%d, offset:0x%X, len:0x%X, last_block:%d",
+			result_code, offset, len, last_block);
+
+	dl->coap.dlc_cc.xfer_res = result_code;
+
+	if ((result_code != COAP_RESPONSE_CODE_OK) && (result_code != COAP_RESPONSE_CODE_CONTENT)) {
+		LOG_ERR("Unexpected response: %*s", len, payload);
+		evt.id = DOWNLOAD_CLIENT_EVT_FRAGMENT;
+		evt.error = result_code;
+	} else {
+		dl->progress += len;
+		LOG_INF("Downloaded %u bytes", dl->progress);
+
+		evt.id = DOWNLOAD_CLIENT_EVT_FRAGMENT;
+		evt.fragment.buf = payload;
+		evt.fragment.len = len;
+	}
+
+	ret = dl->callback(&evt);
+	if (ret) {
+		// todo
+	}
+
+	if (last_block) {
+		LOG_INF("Download complete");
+
+		memset(&evt, 0, sizeof(evt));
+		evt.id = DOWNLOAD_CLIENT_EVT_DONE;
+
+		ret = dl->callback(&evt);
+		if (ret) {
+			// todo
+		}
+	}
+
+	if (last_block || (result_code >= COAP_RESPONSE_CODE_BAD_REQUEST)) {
+		k_sem_give(dl->coap.dlc_cc.xfer_sem);
+	}
+}
+
+#define MAX_RETRIES 5
+int do_coap_client_transfer(struct download_client *const dl)
+{
+	int err;
+	int retry = 0;
+	struct coap_client *cc = dl->coap.dlc_cc.cc;
+
+	struct coap_client_request request = {
+		.method = COAP_METHOD_GET,
+		.confirmable = true, // todo ?
```

Comment:
I think it would be a bad idea to allow folks to do FOTA especially with non-confirmable transfers.

Reply:
is there any reason to expose an option to the user?  this would not always be used for FOTA downloads.